### PR TITLE
User-Defined Grid & 2D Array / Storage Merged v2

### PR DESCRIPTION
**Pull request by EricESeitz**
Closed previous pull request to fix the following:

- Rebased to new master branch (as of 10am 2/13/2019)
- Fixed issue of 2D array not storing unique objects. Now each square holds it's own property (further testing required, test case of 'clickedTimes' successful)

Otherwise identical to previous pull request. In brief:
This is Lauren's and I's separate branches merged and function together. Each square of the user-defined grid corresponds to a coordinate on a 2D array. This array will store the required properties (IsBomb, isClicked, number of bomb neighbors, etc) that we will use to decide outcomes.

I did my best to comment my code and include testing functions for anyone to experiment with. If you have any questions let me know ASAP.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,21 +1,84 @@
...
+  }
+  else {
     validGridSize = true;
   }
 }
 
+//Creates n-dimensional array. Source below
+//https://stackoverflow.com/questions/966225/how-can-i-create-a-two-dimensional-array-in-javascript/966938#966938
+function createArray(length) {
+  var arr = new Array(length || 0),
```
Comment (CompSciLauren) [inline index.js]: Can you change 'var' to 'let' throughout the code?

File: index.js [hunk 2]
```diff
@@ -1,21 +1,84 @@
...
+  for (let i = 0; i < gridSize; i++)
+  {
+    for (let k = 0; k < gridSize; k++)
+    {
+    arr[i][k] = Math.floor(Math.random() * 101); //returns random number between 0 and 100;
+    }
+  }
+
+}
+*/
```
Comment (CompSciLauren) [inline index.js]: Why are lines 36-51 commented out?

Comment (EricESeitz) [reply index.js]: This was the precursor to arrayFiller, used only for testing. Not necessary for the final product but I wasn't sure if anyone wanted to use it for further testing or not. Can be deleted if need be.

File: styles.css [hunk 3]
```diff
@@ -1,5 +1,5 @@
 .square {
-  background-color: white;
+  background-color: lightgray;
```
Comment (CompSciLauren) [inline styles.css]: We want the default state to be white and the clicked state to be lightgray, right?

File: index.js (hunk 1 above)
Comment (EricESeitz) [reply index.js]: I can but could you explain what this would accomplish? I'm not familiar with the difference between the two

File: styles.css (hunk 3 above)
Comment (EricESeitz) [reply styles.css]: Correct, if the square is not a mine or flagged.

I think if the user clicks on a mined square, it should turn red? And flagged squares should be green or a similarly unique color.

File: index.js [hunk 4]
```diff
@@ -34,9 +97,12 @@ $(".square").on("click", function() {
   const elementClicked = $(this);
   const xPos = elementClicked.attr("data-x-coordinate");
   const yPos = elementClicked.attr("data-y-coordinate");
-  handleSquareClick(xPos, yPos);
+  TestFunction(xPos, yPos);
```
Comment (CompSciLauren) [inline index.js]: Why is this changed to TestFunction?

File: index.js [hunk 5]
```diff
@@ -1,21 +1,84 @@
...
+*/
+
+//For storing values in the 2D array, used to create unique instance 'objects'. Can be called on by arr[x][y].<property>
+function squareProperties(isClicked, isBomb, isFlagged, numNeighborMines, testingClickTimes)
+{
+  this.isClicked = isClicked;
+  this.isBomb = isBomb;
+  this.isFlagged = isFlagged;
+  this.numNeighborMines = numNeighborMines;
+  this.testingClickTimes = testingClickTimes;
```
Comment (CompSciLauren) [inline index.js]: Why do we want to track the total number of times a square is clicked? Sorry if we already covered this in a discussion, I can't remember at the moment.

Comment (EricESeitz) [reply index.js]: testingClickTimes? This is a testing property and is not intended to be used in the game. Only to test the functionality of unique 'objects' for each square

File: index.js [hunk 6]
```diff
@@ -34,9 +97,12 @@ $(".square").on("click", function() {
...
 });
 
-function handleSquareClick(xPos, yPos) {
-  console.log(xPos, yPos);
+//testing function for 2D arr, should display (x, y) coordinates on grid click, plus number of times a spesific square has been clicked on
+function TestFunction(x, y)
+{
+  arr[x][y].testingClickTimes = arr[x][y].testingClickTimes + 1;  //increments testingClickTimes up by one for each click
+    alert('Coordinates: (' + x + ', ' + y + ')' + "\nNum Clicked: " + arr[x][y].testingClickTimes);
 }
```
Comment (CompSciLauren) [inline index.js]: Why is this jquery.js file here? JQuery is added to the project already (near the bottom of index.html). The only reason I can think of for adding a JQuery file this way is if you need to be able to work on the project without Internet.

File: index.js (hunk 4 above)
Comment (EricESeitz) [reply index.js]: Correct, TestFunction (as defined below) is merely to test the functionality and could provide a 'basis' for a future function to be built from. If someone needs to come in and follow the logic of the code, IMO these test functions help a lot for getting back up to speed on what's happening and how the logic is flowing.

Test functions are usually weeded out as 'final' code is written to replace them.

File: index.js (hunk 6 above)
Comment (EricESeitz) [reply index.js]: No specific reason other than having it on-hand just in case. You never know what will come up on presentation day. Would it be detrimental to keep it included?

File: index.js [hunk 7]
```diff
@@ -1,21 +1,84 @@
...
+    {
+    arr[i][k] = new squareProperties(0, 0, 0, 0, 0);  //initial assignment, all properties 0
+    }
+  }
+}
+
 drawSquares();
 function drawSquares(square) {
   $("#squareContainer").empty();
+  let id = 0;
```
Comment (CompSciLauren) [inline index.js]: id is never used and should be removed

File: index.js (hunk 2 above)
Comment (CompSciLauren) [reply index.js]: I see. I think we should keep it for now, and when we figure out how to do JS testing we might be able to use it there. I commented below about possibly using Jest for this.

**Comment by CompSciLauren** [comment]
I'm finished reviewing. Code looks good, just had a few questions/comments.

Don't forget to squash your commits before merging! If you click that drop down arrow on "Merge pull request", there should be a Squash & Merge option, or doing it locally works too.

File: index.js (hunk 7 above)
Comment (EricESeitz) [reply index.js]: Removed.

File: styles.css (hunk 3 above)
Comment (EricESeitz) [reply styles.css]: Changed to white

File: index.js (hunk 1 above)
Comment (EricESeitz) [reply index.js]: All var changed to let

**Comment by EricESeitz** [comment]
> 
> 
> I'm finished reviewing. Code looks good, just had a few questions/comments.
> 
> Don't forget to squash your commits before merging! If you click that drop down arrow on "Merge pull request", there should be a Squash & Merge option, or doing it locally works too.

Fixed the colour, removed 'id', changed TestFunction to onClicked for a better descriptive name and changed 'var' to 'let'. 

Thanks for reminding me, definitly need to squash all those comments
